Publicize: Add REST API, follow-up

oskosk opened the pull request:
Follow-up to #10398 addressing some of the comments in https://github.com/Automattic/jetpack/pull/10398#issuecomment-434172435

- [x] Move most of the endpoint code to something like _inc/lib/core-api/wpcom-endpoints/publiciize.php and sync this with WP.com
- [x] Switch to `wpcom/v2` namespace
- [ ] Rename some of the routes. For example, `/publicize/connections` does not just return a list of connections. It runs the connection tests for each one, which is expensive. Currently implemented in this PR: `GET /publicize/services`, `GET /publicize/connections`, `GET /publicize/connection-test-results`.
- [x] Move a post's Publicize Connection info in to the Post endpoint itself via `register_rest_field()`.
- [x] Expose post's Publicize Message info via REST API's meta API.

Everything is up for debate :)

#### Changes proposed in this Pull Request:

All API endpoint code is in `_inc/lib/core-api`.

* `GET /wpcom/v2/publicize/services`
* `GET /wpcom/v2/publicize/connections`
* `GET /wpcom/v2/publicize/connection-test-results`
* `GET /wp/v2/posts/:id` includes `jetpack_publicize_connections` and `meta.jetpack_publicize_message` data.

#### Testing instructions:

##### Automated tests
1. `phpunit --group=publicize`
2. `phpunit --group=rest-api`

##### Manual tests

1. On a connected Jetpack site, with Gutenberg installed.
2. Get to a Gutenberg page (new post)
3. Open the JS console. 
4. Run `wp.apiFetch({path:'/wpcom/v2/publicize/connections' }).then( console.log )` and expect to get a proper list of connections.
5. Run `wp.apiFetch({path:'/wpcom/v2/publicize/services' }).then( console.log )` and expect to get a proper list of compatible social services.
6. Run `wp.apiFetch({path:'/wpcom/v2/publicize/connection-test-results }).then( console.log )` and expect (a slow response containing) a list of connections and whether they are correctly working.
6. Save the post you're currently looking at to get an id. Let's say you get `77`.
5. Run `wp.apiFetch({path:'/wp/v2/posts/77?context=edit' }).then( console.log )` and expect to see the post's Publicize connection details in the `jetpack_publicize_connections` field and the Publicize message string in the `meta.jetpack_publicize_message` field.

#### Proposed changelog entry for your changes:

* None needed

ockham committed:
Copy class-jetpack-publicize-gutenberg.php from #10372 to rest-api.php

ockham committed:
Rename Jetpack_Publicize_Gutenberg to Publicize_REST_API

ockham committed:
Remove obsolete JS globals

ockham committed:
Add Publicize REST API Fields to Post endpoint

ockham committed:
Move Publicize Route Definitions to class.core-rest-api-endpoints.php

ockham committed:
Move process_publicize_from_rest() to Publicize class

ockham committed:
Add tests for process_publicize_from_rest()

ockham committed:
Remove genericons

ockham committed:
Add get_publicize_conns_test_results() function

ockham committed:
Add JSDoc to get_publicize_conns_test_results

mdawaffe committed:
REST API: Publicize: Return `WP_Error` on failure.

This gives us the opportunity to be more informative, and the `WP_Error`
return value is used by all the other `permission_callback`s.

mdawaffe committed:
Typo

oskosk committed:
Move list-connections endpoints to wpcom-endpoints

oskosk committed:
Move list-services endpoint to wpcom-endpoints

oskosk committed:
Move get-post-connections endpoint to wpcom-endpoints

oskosk committed:
Move rest_ functions from publicize.php to endpoint files

oskosk committed:
update to better names for methods

oskosk committed:
address whitespace mess from previous commits

tyxla committed:
Add social logos dependency for now

tyxla committed:
Comment out broken test for now

tyxla committed:
Bump

mdawaffe committed:
Typo

mdawaffe committed:
Publicize: First pass at post fields

mdawaffe commented:
@oskosk, in cce80d96b, I've added a first pass at a possible replacement for the `wpcom/v2/publicize/posts/(?P<post_id>\d+)/connections`: add the data we need to the normal `wp/v2/posts` endpoints.

It doesn't totally work yet :)

mdawaffe committed:
Core API: Generalize `register_rest_field()`

Permission checks,
Type validation,
Context filtering.

Core doesn't do this for us :(

mdawaffe committed:
Todone

mdawaffe committed:
Use Core's REST API Validator

Necessitates switching to an array since Core can't handle `patternProperties`.

mdawaffe committed:
Still not actually working, but the I/O is all set now.

mdawaffe committed:
Publicize API: Convert Services to `WP_REST_Controller`

* A bit more boilerplate, but hopefully more future-proof.
* Adds schema (useful for auto docs)

mdawaffe commented:
> @mdawaffe can you update the testing instructions to include the new changes ?

Yes - I'll get everything written up today. Thanks for the reminder :)

mdawaffe committed:
Publicize: `get_items_permission_check` is not static

mdawaffe committed:
Publicize: Convert Connections to `WP_REST_Controller`

* Connections as now just a list of Connections
* Conneection Test Results is a new `WP_REST_Controller` that does the tests.

mdawaffe committed:
Publicize API: Remove WPCOM_REST_API_V2_Endpoint_Get_Post_Publicize_Connections

The functionality now lives in the Posts endpoint.

mdawaffe committed:
Publicize API: Typo :)

mdawaffe committed:
Publicize API: Add Publicize Message as Post Meta in API

mdawaffe committed:
Publicize API: Oops - meant to move these around before pushing.

mdawaffe committed:
API Field Controller: We only need to return errors from `update_callback`

mdawaffe committed:
Publicize API: Implement Post Connection enable/disable

Needs testing to make sure it works to set the meta after the post update.

mdawaffe commented:
Using `register_meta()` to expose the Publicize Message field works for posts, but Custom Post Types need to register support for `custom-fields` for it to work there. Both Portfolios and Comics register support for `publicize` but not for `custom-fields` so the Publicize Message is currently broken in the API in this PR for those CPTs.

Possible Solutions:
* add `custom-fields` support to the CPTs, or
* don't use `register_meta` - use `register_rest_field()` instead (like this PR does for the Publicize Connections).

mdawaffe committed:
Publicize: Explicitly register `publicize` support for posts.

Rather than making posts an exception, this makes some post_type
stuff easier in the API.

mdawaffe committed:
REST API Fields Controller: Delay registration until rest_api_init.

mdawaffe committed:
REST API Fields Controller: Allow array for `object_type`

mdawaffe committed:
Publicize REST API: Add Connection Data to all Publicize Post Types

mdawaffe committed:
Publicize API: Check ability to publish specific post

mdawaffe committed:
Publicize API: First pass at automated tests

mdawaffe commented:
> > @mdawaffe can you update the testing instructions to include the new changes ?
> 
> Yes - I'll get everything written up today. Thanks for the reminder :)

I lied! I've been working on some automated tests. I'll finish those up tomorrow and actually get it all written up after that.

mdawaffe committed:
Publicize API: Expose Publicize Message Meta for all Post Types

Even post types that don't natively support `custom-fields`.

mdawaffe committed:
REST API Fields Controller: Support `default` value.

Use it when the `get_permission_check()` fails.

mdawaffe committed:
Publicize API: Clean up Permissions

Ability to see/edit Publicize data is determined by `publish_post` not
`edit_post`.

mdawaffe committed:
Publicize API: Clean up Permissions

Rename function to better match what it does.

Some extra cleanup missed in 2e7c1b48f53ce1cb4ba2a02969c0084c10a1a0b6.

mdawaffe committed:
Publicize API: Docs

mdawaffe committed:
Rename files

First part of endpoint plugin filename should be the object it handles.

First part of field plugin filename should be the object it modifies.
Second part the data it adds.

mdawaffe commented:
Resolved. I went with:

> * add `custom-fields` support to the CPTs, or

495ee5dff528396a0ab435bea8c858dd8fab3487

mdawaffe committed:
Publicize API: Remove Post Connection URLs

These were supposed to the URLs corresponding to the content shared on the
Connection's Service.

We don't have that data on the Jetpack side, so let's abandon the idea for
v1.

mdawaffe committed:
Publicize API: More docs

mdawaffe committed:
Publicize API: Remove Unreleased API Code

It's all been translated to other places. Excellent inspiration.

mdawaffe committed:
REST API Fields Controller: Fix Warning in Tests

> PHP Warning:  Declaration of Example_WPCOM_REST_API_V2_Field_Controller::get_permission_check() should be compatible with WPCOM_REST_API_V2_Field_Controller::get_permission_check($object_data, $request) in tests/php/core-api/test_class-wpcom-rest-field-controller.php on line 22

mdawaffe committed:
REST API: Attach to `rest_api_init` in Controllers' Constructors

Go back to `add_action( 'rest_api_init', ... )` in each controller
for better compatibility with how things are loaded on WordPress.com.

mdawaffe committed:
Publicize API: Remove references to `Jetpack_Core_Json_Api_Endpoints`

Use custom error messages.

mdawaffe commented:
With the exception of some tests, I believe D20160-code is up to date with this PR up to 3378fb71f1add6ef625909f328fa830b132accad.

mdawaffe commented:
Technical implementation aside:
* Are we OK with `/publicize/connections`, etc., or should we go with `/publicize-connections`, etc.?
* Do we like the publicize message in meta?  That was my first choice since it's a simple scalar and because I think we should hesitate to add lots of new fields in the main object.

mdawaffe commented:
Note to self: We need to test this back to WordPress 4.8.

tyxla commented on modules/publicize/publicize.php:
```diff
@@ -743,6 +766,80 @@ function blog_id() {
+		 */
+		$capability = apply_filters( 'jetpack_publicize_capability', 'publish_posts' );
+
+		if ( 'publish_posts' == $capability && $post_id ) {
```
Strict equality check here?

mdawaffe replied:
Fixed in cc9d8e9f700906941fdc12c69185d153369f65cb.

tyxla commented on tests/php/core-api/wpcom-fields/test_post-fields-publicize-connections.php:
```diff
@@ -0,0 +1,170 @@
+ * @group publicize
+ * @group rest-api
+ */
+class Test_WPCOM_REST_API_V2_Post_Publicize_Connections_Field extends WP_Test_Jetpack_REST_Testcase {
```
Do you think it makes sense to have some tests for the cases when a user doesn't have the sufficient capabilities?

mdawaffe replied:
Yes - good call. Noted as a TODO in https://github.com/Automattic/jetpack/pull/10605

tyxla reviewed (commented):
This is looking and working really well for me 👍 Amazing work, thanks for laying such solid foundations!

As we've discussed recently, it would probably make sense to split the work to separate PRs to make reviewing and testing easier.

I've spun up a PR for the client side that makes the Gutenberg extension work with it: https://github.com/Automattic/wp-calypso/pull/28463

FWIW, since the latest design changes have been merged, we no longer seem to need the `/publicize/services` endpoint. Also, we don't seem to use `/publicize/connections` at all, but I'm fine with keeping it as a base for the connections tests endpoint, if you think that makes sense.

ockham commented:
Yeah, like @tyxla said:

> As we've discussed recently, it would probably make sense to split the work to separate PRs to make reviewing and testing easier.

Since this is currently ~1600 lines, _and_ based upon my original attempt (#10398), let's maybe move the 'generic' (REST API util) pieces to a fresh PR (based on master)?

mdawaffe commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+class WPCOM_REST_API_V2_Endpoint_List_Publicize_Connection_Test_Results extends WPCOM_REST_API_V2_Endpoint_List_Publicize_Connections {
+	public function __construct() {
+		$this->namespace = 'wpcom/v2';
+		$this->rest_base = 'publicize/connection-test-results';
```
Do you have a suggestion on what to call it?

It's REST, so it has to be a (plural) noun. `connection-tests` sounds ambiguous to me: is it a list of tests that can be performed, or is a list of tests that were performed?

tyxla replied:
You're right, `connection-tests` is indeed ambiguous, so maybe we'd want to keep it as-is.

`test-connections` comes to mind, but it isn't really restful... But then, how about `/publicize/connections/test`? Still ambiguous?

I have no strong feelings, the current one seems clear enough and I'm fine to go with it if you like it.

mdawaffe replied:
I like the explicitness of the current name, so let's stick with it :)

mdawaffe committed:
REST Fields Controller: More explicit function name

mdawaffe committed:
REST Fields Controller: Code Standards

mdawaffe committed:
REST Fields Controller: PHPDoc

mdawaffe committed:
REST Fields Controller: Clarify odd WP_Error used in recursion.

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,308 @@
+	 *
+	 * @param array $schema
+	 */
+	final public function default_value( $schema ) {
```
Fixed in 4f900bbd9.

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,308 @@
+
+		// If you have something more complicated, use $schema['default'];
+		switch ( isset( $schema['type'] ) ? $schema['type'] : 'null' ) {
+		case 'string' :
```
Huh - https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation

I thought the code standard was to not use tabs here.

Fixed in 5ade7a41d.

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,308 @@
+	 * @param mixed $object_data Probably a WordPress object (e.g., WP_Post)
+	 * @param string $field_name Should always match `->field_name`
+	 * @param WP_REST_Request
+	 * @param $object_type Should always match `->object_type`
```
Thanks. Cleaned up in ac026b3e1.

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,308 @@
+		$permission_check = $this->update_permission_check( $value, $object_data, $request );
+
+		if ( ! $permission_check ) {
+			return;
```
After reading your comment, I thought the answer was "yes". But I looked more into the Core code this talks to, and I think it's an implementation error to return false here.

I added some warning in ac026b3e1. We can make it return an error instead of the warning, though, if we think an error is better.

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,308 @@
+	 */
+	final function filter_response_by_context( $value, $schema, $context ) {
+		if ( ! $this->is_valid_for_context( $schema, $context ) ) {
+			return new WP_Error( '__wrong-context__' );
```
Clarified in 3a427bb0b. The odd error code is intentional :) It's designed to let you know that something odd is going on. I guess it worked ;)

ockham commented on _inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php:
```diff
@@ -0,0 +1,256 @@
+	 * Update the connections slated to be shared to.
+	 *
+	 * @param array $requested_connections
+	 *              Items are eitheer `{ id: (string) }` or `{ service_name: (string) }`
```
Typo: `eitheer`

mdawaffe replied:
Fixed in ab85f6ebd.

mdawaffe commented:
> Since this is currently ~1600 lines, _and_ based upon my original attempt (#10398), let's maybe move the 'generic' (REST API util) pieces to a fresh PR (based on master)?

Done in #10600, which I think addresses all the feedback so far about the Field Controller. I'll rebase this one (or create a new one) for the Publicize stuff later.

mdawaffe commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+					'id' => $publicize->get_connection_unique_id( $connection ),
+					'service_name' => $service_name,
+					'display_name' => $publicize->get_display_name( $service_name, $connection ),
+					'global' => 0 == $connection_data['user_id'],
```
I left the loose check in on purpose. I worry that we've stored `'0'` for some sites and `0` for others.

mdawaffe commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+require_once dirname( __FILE__ ) . '/publicize-connections.php';
+
+/**
+ * Publicize: List Conection Test Result Data
```
Fixed in ab85f6ebd.

mdawaffe commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+		}
+
+		$response = rest_ensure_response( $items );
+                $response->header( 'X-WP-Total', count( $items ) );
```
Thanks! Fixed in 46f9f7a73.

mdawaffe commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+		$this->namespace = 'wpcom/v2';
+		$this->rest_base = 'publicize/connection-test-results';
+
+		add_action( 'rest_api_init', array( $this, 'register_routes' ) );
```
I'd like to do what you suggest. Let's leave it for a later refactor, though, since doing so would require a bunch of changes on WordPress.com, too. All the endpoints there follow this pattern, and some of the loader code there expects it.

mdawaffe commented:
I've addressed most of the feedback on this PR (other than a couple style points, which we can discuss on the new PRs).

Let's close this and move the discussion to:
* #10600 - REST API Field Controller
* #10605 - Publicize API

Thanks for the discussions so far!
